### DISCUSSION TITLE
can you add ML.NET to this library

### DISCUSSION THREAD

**Discussion by tiger2014** [DaveSkender/Stock.Indicators#1176] [Ideas]
Hi there,

Can you add ML.NET to do some stock price prediction for teaching purposes? For example, according to MACD and pre-marked stock price points for rise and fall, use ML.NET to predict price trend.

Thanks

**Comment by DaveSkender** [comment]
You can certainly use this package in an ML project that would predict price trends.

As far as adding it inside the package: if we were to venture into the analysis and strategy development world, I'd likely bundle it in an entirely separate project, since we like maintain a [single responsibility](https://github.com/DaveSkender/Stock.Indicators/discussions/648) on this library (e.g. quotes in > raw indicators out).

There's certainly been some interests for analysis and back testing:

- #822
- #484
- #320

Including some of your nice recommendations:
- #874
- #809

With that said, I'm open to exploring what this separate project might look like.  Any interest in contributing some more thoughts and details, or some high level user stories/use cases?

**Reply by tiger2014** [reply]
I have two ideas. The first idea is that I will mark all these trend starting points manually and add the indicators that we consider important through vector analysis. The values of these indicators will be used as vectors for analysis. Then, through this vector analysis, when a new candlestick appears, I will recalculate this vector and search for the most matching result in the historical data. From then on, I will calculate the possibility of its rise and fall using this method.

The second approach is to use calculations to determine the starting points. For example, we can establish a rule: if we enter at a certain point and hit the stop-loss, it does not count as a starting point. However, if after entering, it doesn't reach our stop-loss and continues to rise/down, we can set a rule such as a 1:2 risk-reward ratio. When the loss ratio reaches double the initial risk, we consider it a valid starting point for the upward trend. We use calculations to mark these starting points, and the calculation of subsequent indicators is the same as before.

Because I lack professional knowledge in ML, this idea seems very basic to professionals, but it was summed up by me in my trading.

And the key point is that I don't know how to vectorize (i.e. feature engineering). If his done, we can let ML.NET to train the model.

I holp someone can share his specific methods of vectorization.

**Reply by DaveSkender** [reply]
I'm not an ML expert either ... and there's very few people who'd say they are.  From what I understand, vectors are mathematical representations of _memory_ that'd allow a machine algorithm to find and process information faster and with more accuracy than a rote scan-type of search.

This is a rather large leap into the machine learning realm.  Sounds like fun, but definitely something to build separately.  There's some really good free online courses from Microsoft to help you [get started with ML.NET](https://dotnet.microsoft.com/en-us/apps/machinelearning-ai/ml-dotnet) if that's where you are in your journey.  I've gone through quite a bit of it and it's quite easy to start building your own AI models through a process of training the model (and continuous learning to improve it).  Might be fun to pick that up again.

- ref: [what is a vector database?](https://www.cloudflare.com/learning/ai/what-is-vector-database)

**Reply by tiger2014** [reply]
Thanks!

The results trained using ML.NET (AutoML) have a very low prediction accuracy.
